src/main/java/bio/terra/stairway/DbRetry.java

```diff
@@ -0,0 +1,69 @@
+package bio.terra.stairway;
+
+import bio.terra.stairway.exception.DatabaseOperationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.SQLException;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The DbRetry class provides a common retry framework for the DAOs. A new DbRetry class is
+ * constructed at runtime for use wrapping inner database calls. The retry algorithm is random
+ * backoff with a maximum number of retries. At this time, there is no use case for setting
+ * different retry times for different calls. If there is, then an additional constructor can be
+ * added to allow specification of what are coded as statics.
+ */
+public class DbRetry {
+  private static final Logger logger = LoggerFactory.getLogger(DbRetry.class);
+  private static final int MAX_RETRIES = 20;
+  private static final int WAIT_MIN_MS = 250;
+  private static final int WAIT_MAX_MS = 1000;
+  private static final String PSQL_SERIALIZATION_FAILURE = "40001";
```

Comment:
One of the exception I saw from log is 
`org.postgresql.util.PSQLException: An I/O error occurred while sending to the backend.` [[log]](org.postgresql.util.PSQLException: An I/O error occurred while sending to the backend.)
I didn't find the number for that exception. Does it covered by any of those code?

Reply:
I believe that will be a SQLstate beginning with PSQL_CONNECTION_ISSUE_PREFIX ("08").
The SQLstate is a member of the PSQLException object, so if you have logging that prints the whole object, you should have that value.